src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1830,12 +1830,46 @@ private static BoundExpression SkipReferenceConversions(BoundExpression possibly
             return possiblyConversion;
         }
 
-        // https://github.com/dotnet/roslyn/issues/30140: Track nullable state from ??=.
         public override BoundNode VisitNullCoalescingAssignmentOperator(BoundNullCoalescingAssignmentOperator node)
         {
-            var result = base.VisitNullCoalescingAssignmentOperator(node);
-            SetResult(node);
-            return result;
+
+            BoundExpression leftOperand = node.LeftOperand;
+            BoundExpression rightOperand = node.RightOperand;
+
+            TypeSymbolWithAnnotations leftResult = VisitRvalueWithResult(leftOperand);
+
+            if (leftResult.ValueCanBeNull() == false)
+            {
+                ReportNonSafetyDiagnostic(ErrorCode.HDN_ExpressionIsProbablyNeverNull, leftOperand.Syntax);
+            }
+
+            var leftState = this.State.Clone();
+
+            int leftSlot = MakeSlot(leftOperand);
+            if (leftSlot != -1)
+            {
+                leftState[MakeSlot(leftOperand)] = NullableAnnotation.NotNullable;
+            }
+
+            if (leftResult.NullableAnnotation.IsAnyNotNullable())
```

Comment:
>leftResult.NullableAnnotation.IsAnyNotNullable() [](start = 16, length = 48)

It doesn't look like this test is going to properly handle unconstrained type parameters.